native/core/src/execution/fair_memory_pool.rs

```diff
@@ -0,0 +1,159 @@
+        f.debug_struct("CometFairMemoryPool")
+            .field("pool_size", &self.pool_size)
+            .field("used", &state.used)
+            .field("num", &state.num)
+            .finish()
+    }
+}
+
+impl CometFairMemoryPool {
+    pub fn new(
+        task_memory_manager_handle: Arc<GlobalRef>,
+        pool_size: usize,
+    ) -> CometFairMemoryPool {
+        Self {
+            task_memory_manager_handle,
+            pool_size,
+            state: Mutex::new(CometFairPoolState { used: 0, num: 0 }),
+        }
+    }
+
+    fn acquire(&self, additional: usize) -> CometResult<i64> {
+        let mut env = JVMClasses::get_env()?;
+        let handle = self.task_memory_manager_handle.as_obj();
+        unsafe {
+            jni_call!(&mut env,
+              comet_task_memory_manager(handle).acquire_memory(additional as i64) -> i64)
+        }
+    }
+
+    fn release(&self, size: usize) -> CometResult<()> {
+        let mut env = JVMClasses::get_env()?;
+        let handle = self.task_memory_manager_handle.as_obj();
+        unsafe {
+            jni_call!(&mut env, comet_task_memory_manager(handle).release_memory(size as i64) -> ())
+        }
+    }
+}
+
+unsafe impl Send for CometFairMemoryPool {}
+unsafe impl Sync for CometFairMemoryPool {}
+
+impl MemoryPool for CometFairMemoryPool {
+    fn register(&self, _: &MemoryConsumer) {
+        let mut state = self.state.lock();
+        state.num = state.num.checked_add(1).unwrap();
+    }
+
+    fn unregister(&self, _: &MemoryConsumer) {
+        let mut state = self.state.lock();
+        state.num = state.num.checked_sub(1).unwrap();
+    }
+
+    fn grow(&self, reservation: &MemoryReservation, additional: usize) {
+        self.try_grow(reservation, additional).unwrap();
+    }
+
+    fn shrink(&self, reservation: &MemoryReservation, subtractive: usize) {
+        if subtractive > 0 {
+            let mut state = self.state.lock();
+            let size = reservation.size();
+            if size < subtractive {
+                panic!("Failed to release {subtractive} bytes where only {size} bytes reserved")
+            }
+            self.release(subtractive)
+                .unwrap_or_else(|_| panic!("Failed to release {} bytes", subtractive));
+            state.used = state.used.checked_sub(subtractive).unwrap();
+        }
+    }
+
+    fn try_grow(
+        &self,
+        reservation: &MemoryReservation,
+        additional: usize,
+    ) -> Result<(), DataFusionError> {
+        if additional > 0 {
+            let mut state = self.state.lock();
+            let num = state.num;
+            let limit = self.pool_size.checked_div(num).unwrap();
+            let size = reservation.size();
+            if limit < size + additional {
+                return Err(resources_datafusion_err!(
+                    "Failed to acquire {additional} bytes where {size} bytes already reserved and the fair limit is {limit} bytes, {num} registered"
+                ));
```

Comment:
```suggestion
                return resources_err!(
                    "Failed to acquire {additional} bytes where {size} bytes already reserved and the fair limit is {limit} bytes, {num} registered"
                );
```

Reply:
Looks like `resources_err` creates backtrace? Actually this error is expected to happen many times for triggering the memory consumer to spill. Not sure if we want backtrace everytime

Reply:
resources_datafusion_err also has the optional backtrace.

`resources_err` is the shorthand for `Err(resources_datafusion_err!)`

And backtrace will be enabled only if DF is built with `backtrace` feature, usually it is needed only for debugging or investigation